software/controller/lib/hal/uart_dma.h

```diff
@@ -15,96 +15,70 @@ limitations under the License.
-  virtual void OnRxError(RxError) = 0;
-};
 
-class UartDmaTxListener {
- public:
-  // Called on DMA TX complete
-  virtual void OnTxComplete() = 0;
-  // Called on TX errors
-  virtual void OnTxError() = 0;
-};
+#include "hal_stm32_regs.h"
+#include "serial_listeners.h"
 
-class DMACtrl {
+class DmaCtrl {
  public:
-  explicit DMACtrl(DmaReg *const dma) : dma_(dma) {}
-  void Init() {
+  explicit DmaCtrl(DmaReg *const dma) : dma_(dma) {}
+  void init() {
     // UART3 reception happens on DMA1 channel 3
     dma_->channel_select.c3s = 0b0010;
     // UART3 transmission happens on DMA1 channel 2
     dma_->channel_select.c2s = 0b0010;
   }
 
  private:
-  DmaReg *const dma_;
+  DmaReg *const dma_{nullptr};
 };
 
 class UartDma {
  public:
-  UartDma(UartReg *const uart, DmaReg *const dma, uint8_t tx_chan, uint8_t rx_chan,
-          UartDmaRxListener *rxl, UartDmaTxListener *txl, char match_char)
+#ifdef TEST_MODE
+  UartDma() = default;
+#endif
+  UartDma(UartReg *const uart, DmaReg *const dma, uint8_t tx_channel, uint8_t rx_channel,
+          char match_char)
       : uart_(uart),
         dma_(dma),
-        tx_channel_(tx_chan),
-        rx_channel_(rx_chan),
-        rx_listener_(rxl),
-        tx_listener_(txl),
+        tx_channel_(tx_channel),
+        rx_channel_(rx_channel),
         match_char_(match_char) {}
 
-  void Init(int baud);
-  // Returns true if DMA TX is in progress
-  bool TxInProgress() const;
-  // Returns true if DMA RX is in progress
-  bool RxInProgress() const;
+  void init(uint32_t baud);
 
-  // Sets up UART3 to transfer [length] characters from [buf]
-  // Returns false if DMA transmission is in progress, does not
-  // interrupt previous transmission.
-  // Returns true if no transmission is in progress
-  bool StartTX(char *buf, uint32_t length);
+  [[nodiscard]] bool start_tx(uint8_t *buf, uint32_t length, TxListener *txl);
+  [[nodiscard]] bool start_rx(uint8_t *buf, uint32_t length, RxListener *rxl);
 
-  uint32_t GetRxBytesLeft();
+  bool tx_in_progress() const;
+  bool rx_in_progress() const;
 
-  void StopTX();
+  void stop_tx();
+  void stop_rx();
 
-  // Sets up reception of at least [length] chars from UART3 into [buf]
-  // [timeout] is the number of baudrate bits for which RX line is
-  // allowed to be idle before asserting timeout error.
-  // Returns false if reception is in progress, new reception is not
-  // setup. Returns true if no reception is in progress and new reception
-  // was setup.
+  uint32_t rx_bytes_left();
 
-  bool StartRX(char *buf, uint32_t length, uint32_t timeout);
-  void StopRX();
-  void CharMatchEnable();
+  void enable_character_match();
 
-  void UartISR();
-  void DmaRxISR();
-  void DmaTxISR();
+  void UART_interrupt_handler();
+  void DMA_rx_interrupt_handler();
+  void DMA_tx_interrupt_handler();
 
  private:
-  UartReg *const uart_;
-  DmaReg *const dma_;
+  typedef struct {
+    uint32_t ignored : 8;
+    uint32_t bits : 24;
+  } uint24_t;
```

Comment:
I don't see this type used anywhere, am I missing something?

Reply:
Ah, yes, this is true. There was a function called `DurationToBits` and was used for `rx_timeout`, but since this commit:
https://github.com/RespiraWorks/Ventilator/pull/1136/commits/8120e642e13da219b1113e53b088c7a1adc95619
This functionality was removed, so I removed the unused function as well.
I can probably remove this struct, but I am not sure if we may ever want to actually use the timeout feature, we may want to keep it around?

Reply:
How, about this, I will add the code for this struct and function here in the comments. If we ever do need it, we can, with some effort track down this PR and resurrect the code. At least it will not get lost in the squashing.